From 3b75a77985deab6c9161846e8674ad45ac4cdcf4 Mon Sep 17 00:00:00 2001
From: Carlos Garnacho <carlosg@gnome.org>
Date: Mon, 7 Jul 2025 17:25:44 +0200
Subject: [PATCH 1/4] media-keys: Use GAppLaunchContext directly

The only advantage we might ever obtain from GDK here is integrating
with startup notification for applications being launcher. However,
we do just receive a timestamp, which is by itself not sufficient to
integrate with startup notification on Wayland.

While we think that through, using GTK has other side effects, so
it's better to avoid GdkAppLaunchContext here altogether.
---
 plugins/media-keys/gsd-media-keys-manager.c | 21 ++++++++-------------
 plugins/media-keys/main.c                   |  5 -----
 plugins/media-keys/meson.build              |  1 -
 3 files changed, 8 insertions(+), 19 deletions(-)

diff --git a/plugins/media-keys/gsd-media-keys-manager.c b/plugins/media-keys/gsd-media-keys-manager.c
index 2811114b..660a8ffc 100644
--- a/plugins/media-keys/gsd-media-keys-manager.c
+++ b/plugins/media-keys/gsd-media-keys-manager.c
@@ -34,7 +34,6 @@
 #include <glib.h>
 #include <glib/gi18n.h>
 #include <gio/gio.h>
-#include <gdk/gdk.h>
 #include <gio/gdesktopappinfo.h>
 #include <gio/gunixfdlist.h>
 
@@ -1001,13 +1000,12 @@ launch_app (GsdMediaKeysManager *manager,
 	    GAppInfo            *app_info,
 	    gint64               timestamp)
 {
-	GError *error = NULL;
-        GdkAppLaunchContext *launch_context;
+        g_autoptr (GError) error = NULL;
+        g_autoptr (GAppLaunchContext) launch_context = NULL;
 
         /* setup the launch context so the startup notification is correct */
-        launch_context = gdk_display_get_app_launch_context (gdk_display_get_default ());
-        gdk_app_launch_context_set_timestamp (launch_context, timestamp);
-        set_launch_context_env (manager, G_APP_LAUNCH_CONTEXT (launch_context));
+        launch_context = g_app_launch_context_new ();
+        set_launch_context_env (manager, launch_context);
 
         g_signal_connect_object (launch_context,
                                  "launched",
@@ -1019,9 +1017,7 @@ launch_app (GsdMediaKeysManager *manager,
 		g_warning ("Could not launch '%s': %s",
 			   g_app_info_get_commandline (app_info),
 			   error->message);
-		g_error_free (error);
 	}
-        g_object_unref (launch_context);
 }
 
 static void
@@ -1279,7 +1275,7 @@ static void
 do_home_key_action (GsdMediaKeysManager *manager,
 		    gint64               timestamp)
 {
-	GdkAppLaunchContext *launch_context;
+	g_autoptr (GAppLaunchContext) launch_context = NULL;
 	g_autoptr (GFile) file = NULL;
 	g_autoptr (GError) error = NULL;
 	g_autofree char *uri;
@@ -1287,10 +1283,9 @@ do_home_key_action (GsdMediaKeysManager *manager,
 	file = g_file_new_for_path (g_get_home_dir ());
 	uri = g_file_get_uri (file);
 
-	launch_context = gdk_display_get_app_launch_context (gdk_display_get_default ());
-	gdk_app_launch_context_set_timestamp (launch_context, timestamp);
+	launch_context = g_app_launch_context_new ();
 
-	if (!g_app_info_launch_default_for_uri (uri, G_APP_LAUNCH_CONTEXT (launch_context), &error))
+	if (!g_app_info_launch_default_for_uri (uri, launch_context, &error))
 		g_warning ("Failed to launch '%s': %s", uri, error->message);
 }
 
@@ -2578,7 +2573,7 @@ on_accelerator_activated (ShellKeyGrabber     *grabber,
         if (!g_variant_dict_lookup (&dict, "device-node", "s", &device_node))
               device_node = NULL;
         if (!g_variant_dict_lookup (&dict, "timestamp", "u", &timestamp))
-              timestamp = GDK_CURRENT_TIME;
+              timestamp = 0L;
         if (!g_variant_dict_lookup (&dict, "action-mode", "u", &mode))
               mode = 0;
 
diff --git a/plugins/media-keys/main.c b/plugins/media-keys/main.c
index 52ac0ea8..e2de6c05 100644
--- a/plugins/media-keys/main.c
+++ b/plugins/media-keys/main.c
@@ -1,13 +1,8 @@
-#include <gtk/gtk.h>
-
 #include "gsd-main-helper.h"
 #include "gsd-media-keys-manager.h"
 
 int
 main (int argc, char **argv)
 {
-        gtk_init ();
-
         return gsd_main_helper (GSD_TYPE_MEDIA_KEYS_MANAGER, argc, argv);
 }
-
diff --git a/plugins/media-keys/meson.build b/plugins/media-keys/meson.build
index 2d95686c..5bd3f763 100644
--- a/plugins/media-keys/meson.build
+++ b/plugins/media-keys/meson.build
@@ -29,7 +29,6 @@ deps = plugins_deps + [
   gsettings_desktop_dep,
   libcommon_dep,
   gnome_desktop_dep,
-  gtk_dep,
   libgvc_dep,
   libpulse_mainloop_glib_dep,
   m_dep,
-- 
2.49.0

